### ISSUE TITLE
Stuck at finish line

### ISSUE THREAD

**Issue by bugsyb** [nextcloud/recognize#562]
### Which version of recognize are you using?

latest

### Enabled Modes

Object recognition, Face recognition, Video recognition, Music recognition

### TensorFlow mode

GPU mode

### Which Nextcloud version do you have installed?

25.0.2.3

### Which Operating system do you have installed?

Debian 11

### Which Docker container are you using to run Nextcloud? (if applicable)

v20.10.12

### How much RAM does your server have?

32GB

### What processor Architecture does your CPU have?

amd64

### Describe the Bug

Processing of all files is at the last line and got stuck:
```
SELECT COUNT(*) FROM oc_recognize_queue_imagenet; SELECT COUNT(*) FROM oc_recognize_queue_faces; SELECT COUNT(*) FROM oc_recognize_queue_landmarks;
 count queue ImageNet:  323
 count queue faces:  144
 count queue landmarks: 0

-------------------------------------------------------------------------------------------------------------------------
SELECT (last_run,last_checked,reserved_at,execution_duration,argument,class) FROM public.oc_jobs WHERE class like '%Recognize%' ORDER BY reserved_at, argument;

 (1667293808,1671044049,0,620,null,"OCA\\Recognize\\BackgroundJobs\\ClassifyImagesJob")
 (1667292916,1671044049,0,635,null,"OCA\\Recognize\\BackgroundJobs\\ClassifyAudioJob")
 (1671032753,1671032829,0,8,"{""storageId"":10,""rootId"":31946}","OCA\\Recognize\\BackgroundJobs\\ClassifyImagenetJob")
(3 rows)

```

Interestingly, when the processing was happening earlier, faces queue for last hours was always going down only to 144 (never below).

In paralle, could be related, but not necessarily the culprit on the PostgresDB side observed this:
```
15:58:12.130 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.130 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559985, 99) already exists.
15:58:12.130 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.138 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.138 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559985, 100) already exists.
15:58:12.138 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.142 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.142 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559985, 67) already exists.
15:58:12.142 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.502 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.502 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559992, 98) already exists.
15:58:12.502 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.506 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.506 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559992, 100) already exists.
15:58:12.506 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.513 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.513 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3559992, 67) already exists.
15:58:12.513 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.942 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.942 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560031, 99) already exists.
15:58:12.942 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.948 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.948 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560031, 100) already exists.
15:58:12.948 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:12.952 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:12.952 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560031, 67) already exists.
15:58:12.952 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.307 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.307 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560101, 148) already exists.
15:58:13.307 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.311 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.311 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560101, 100) already exists.
15:58:13.311 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.315 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.315 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560101, 67) already exists.
15:58:13.315 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.705 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.705 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560110, 99) already exists.
15:58:13.705 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.709 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.709 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560110, 100) already exists.
15:58:13.709 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
15:58:13.713 CET [4222] ERROR:  duplicate key value violates unique constraint "oc_systemtag_object_mapping_pkey"
15:58:13.713 CET [4222] DETAIL:  Key (objecttype, objectid, systemtagid)=(files, 3560110, 67) already exists.
15:58:13.713 CET [4222] STATEMENT:  INSERT INTO "oc_systemtag_object_mapping" ("objectid", "objecttype", "systemtagid") VALUES($1, $2, $3)
```

### Expected Behavior

Finish processing of all files

### To Reproduce

Unknwon

### Debug log

Provided above.

**Comment by marcelklehr** [comment]
Anything in the nextcloud logs?

**Comment by bugsyb** [comment]
Nothing - literally (see attached).
[nc2.log](https://github.com/nextcloud/recognize/files/10230301/nc2.log)

**Comment by bugsyb** [comment]
For tests - I've removed offending entries from `oc_systemtag_object_mapping`. These got recreated and same errors occurred again. "Automatically" am back to square one.

**Comment by bugsyb** [comment]
Further efforts to debug led me to truncating oc_recognize_queue_imagenet and injecting first of records there (table dump and then insert of one of records).
Even that single file hasn't been processed. It has been confirmed that file exist, is readable. 

Switching on and off landmarks, 5 new files were added to imagenet queue and got processed, but not all.
Again two files to be processed stayed there. These were relatively new files from yesterday for sure were present on the file system and accessible fully via Nextcloud interface, both preview as well as full files.
In parallel full filesystem scan was done for accounts in question to make sure nothing is missed.

At the moment I did disable image object recognition and re-enabled it. To my surprise queue size did grow to 5k and seems like might be growing further.
Feeling is that disabling and re-enabling causes Recognize to re-crawl from the start through all files.

To further debug this, the best would be an ability to manually run processing of selected queue in manual mode with debug messages enabled to see what fails and hopefully why.

**Comment by marcelklehr** [comment]
> Feeling is that disabling and re-enabling causes Recognize to re-crawl from the start through all files.

Yes, that is the case, although files will be removed from the queue again, if they've been processed already.

**Comment by marcelklehr** [comment]
> Nothing - literally (see attached).

No error messages in nextcloud logs is curious.

**Comment by bugsyb** [comment]
Literally no errors, nothing - the earlier attached file shows full dump of logs from single cron run.
After re-enabling as system goes through all files again, there are plenty of these dup errors from Postgres.
Therefore this is also cul de sac.

I'm clueless now and feeling like best will be to manually run the process on item by item with full debug on.
It seems like it didn't even get to run TensorFlow and stops earlier. Maybe at preview generation? To check if there's no problem with that, I looked at physical existence of files and all looks good - files are in there as per earlier comment and accessible via NC as well as directly file system level.

How would I go about to run image net manually, i.e. for a single item from the queue?

**Comment by bugsyb** [comment]
Found answer, but not the root cause why it did happen, this thing is deeper in the code/race condition effect?

It's down to the fact that there were no Jobs defined for the Storage_ID of files left in the queue.
The job was for storage#10, whilst files were mix of other than storage#10.

Not sure how I did end up like this and have no reproduction so far, though would be good to have fail-safe mechanism to catch such situation should it happen.

**Comment by bugsyb** [comment]
Got into same situation again though more interesting this time.
Image queue, shows 12 images left to process, all with same storage ID#10, there's job in Jobs to be executed for this storageID
```
 ("{""storageId"":10,""rootId"":31946}",1671362822,1671363002,0,13,"OCA\\Recognize\\BackgroundJobs\\ClassifyImagenetJob")
 (null,1667293808,1671390602,0,620,"OCA\\Recognize\\BackgroundJobs\\ClassifyImagesJob")
 (null,1667292916,1671390603,0,635,"OCA\\Recognize\\BackgroundJobs\\ClassifyAudioJob")
(3 rows)
```
There is nothing in logs, though no debug enabled - happy to do so.

More interesting is that this time, there are 144 items in faces queue and no Job to have them processed.
All of these in faces queue are from other storage ID than #10. 

Feeling is that there's a glitch in catching these situations.

One more item, which might help catching other bug - there were 6 entries of which
- 3 entries didn't have existing storage ID anymore.
- 3 entries didn't have existing files anymore (based on oc_filecache);

These were caught with checking entries as below:
```
SELECT q.id, q.file_id, q.storage_id, s.id, f.path
FROM oc_recognize_queue_faces q LEFT JOIN oc_filecache f ON q.file_id  = f.fileid LEFT JOIN oc_storages s ON q.storage_id = s.numeric_id 
ORDER BY  q.id;
```

And then respectively deleting them after confirming with below:
```
SELECT q.id, q.file_id, q.storage_id FROM oc_recognize_queue_faces q 
WHERE q.file_id NOT IN (SELECT f.fileid FROM oc_filecache f);
DELETE FROM oc_recognize_queue_faces q 
WHERE q.file_id NOT IN (SELECT f.fileid FROM oc_filecache f);
```

```
SELECT q.id, q.file_id, q.storage_id FROM oc_recognize_queue_faces q 
WHERE q.storage_id NOT IN (SELECT s.numeric_id FROM oc_storages s);
DELETE FROM oc_recognize_queue_faces q 
WHERE q.storage_id NOT IN (SELECT s.numeric_id FROM oc_storages s);
```
This didn't unlock processing, understanding is that it's due to missing Job (this is a separate issue from stuck ImageNet queue processing, am just not sure if a separate issue ticket is preferred - let me know).

**Comment by marcelklehr** [comment]
> More interesting is that this time, there are 144 items in faces queue and no Job to have them processed.

Mmh, weird. not sure where this goes wrong.

> 3 entries didn't have existing files anymore (based on oc_filecache);

These should be fine, as when the classifier stumbles upon them it will just remove them from the queue and continue.

**Comment by bugsyb** [comment]
Given I did run into the situation more than once - would you have some suggestions on how to debug it further or at least un-lock?
I'm literally stuck.

**Comment by marcelklehr** [comment]
@bugsyb Sorry for the wait, I've been stuck on other tasks and didn't really have an idea about this issue.

You should be able to unstuck by creating a faces job on oc_jobs: \OCA\Recognize\BackgroundJobs\ClassifyFacesJob with argument `{"storageId": XXX, "rootId": XXX}` (you can get the ids from the queue table)


Only instances of the job getting removed in the code are

- Not classifying files of storage '.$storageId. ' using '.$model. ' because model is disabled
- Removing \OCA\Recognize\BackgroundJobs\ClassifyFacesJob with argument ... from oc_jobs
  - happens because no items are left in the queue. Adding items to the queue will create a new job.
